File: backend/models/schema.py
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any, Union

class LoreEntry(BaseModel):
    id: str
    name: str
    content: str
    keywords: List[str]

class TrackedItem(BaseModel):
    id: str
    name: str
    positionInList: int
    dataType: str
    visibility: str
    description: str
    updateInstructions: str
    initialValue: str
    initialValueBasedOnPC: str
    autoUpdate: bool

class TriggerCondition(BaseModel):
    id: str
    data: str
    type: str
    category: str

class TriggerEffect(BaseModel):
    id: str
    data: Union[str, Dict[str, Any]]
    type: str

class TriggerEvent(BaseModel):
    id: str
    name: str
    triggerEffects: List[TriggerEffect]
    triggerConditions: List[TriggerCondition]

class InstructionBlock(BaseModel):
    id: str
    name: str
    content: str

class Character(BaseModel):
    name: str
    description: str
    characterId: str
    skills: Dict[str, int]
    initialTrackedItemValues: List[Any] = []

class WorldSchema(BaseModel):
    title: str
    description: str
    background: str
    instructions: str
    authorStyle: str
    recommendedAIModel: str
    objective: str
    skills: List[str]
    possibleCharacters: List[Character]
    triggerEvents: List[TriggerEvent]
    instructionBlocks: List[InstructionBlock]
    loreBookEntries: List[LoreEntry]
    trackedItems: List[TrackedItem]
    imageModel: Optional[str] = None
    imageStyleCharacterPre: Optional[str] = None
    imageStyleCharacterPost: Optional[str] = None
    imageStyleNonCharacterPre: Optional[str] = None
    imageStyleNonCharacterPost: Optional[str] = None

File: frontend/src/types/schema.ts
export interface WorldSchema {
  title: string;
  description: string;
  background: string;
  instructions: string;
  authorStyle: string;
  firstInput: string;
  objective: string;
  possibleCharacters: Character[];
  triggerEvents: TriggerEvent[];
  instructionBlocks: InstructionBlock[];
  loreBookEntries: LoreBookEntry[];
  trackedItems: TrackedItem[];
  victoryCondition: GameCondition;
  defeatCondition: GameCondition;
  imageStyleCharacterPre?: string;
  imageStyleCharacterPost?: string;
  imageStyleNonCharacterPre?: string;
  imageStyleNonCharacterPost?: string;
}

export interface Character {
  characterId: string;
  name: string;
  description: string;
  portrait: string;
  skills: Record<string, number>;
}

export interface TriggerEvent {
  id: string;
  name: string;
  triggerConditions: TriggerCondition[];
  triggerEffects: TriggerEffect[];
}

export interface TriggerCondition {
  id: string;
  type: "triggerOnEvent" | "valueCheck";
  data: any;
}

export interface TriggerEffect {
  id: string;
  type: "effectShowMessage" | "effectSetTrackedItemValue" | "effectModifyInstructionBlock";
  data: any;
}

export interface InstructionBlock {
  id: string;
  name: string;
  content: string;
}

export interface LoreBookEntry {
  id: string;
  name: string;
  content: string;
  keywords: string[];
}

export interface TrackedItem {
  id: string;
  name: string;
  dataType: "text" | "number" | "xml";
  initialValue: any;
  updateInstructions: string;
  autoUpdate: boolean;
}

export interface GameCondition {
  condition: string;
  text: string;
  alreadyFired: boolean;
}

export interface GameState {
  world: WorldSchema;
  session: {
    character: Character;
    currentValues: Record<string, any>;
    modifiedInstructions: Record<string, string>;
    firedTriggers: string[];
    history: { role: "user" | "model"; content: string; narrative?: string; images?: string[] }[];
  };
}
